Strange Error on array literal as call argument where ReadSeq[T] is expected

When compiling ponycheck against ponyc 0.19.1 i am running into a strange error message:

```
/home/travis/build/mfelsche/ponycheck/ponycheck/generator.pony:832:24: array element not a subtype of specified array type
          (range_1_size, range_1)
                       ^
    Info:
    /home/travis/build/mfelsche/ponycheck/ponycheck/generator.pony:245:34: array type: (USize val, this->Generator[U32 val] box)
        weighted_generators: ReadSeq[WeightedGenerator[T]])
                                     ^
    /home/travis/build/mfelsche/ponycheck/ponycheck/generator.pony:832:24: element type: (USize val, Generator[U32 val] box)
              (range_1_size, range_1)
                           ^
    /home/travis/build/mfelsche/ponycheck/ponycheck/generator.pony:423:7: Generator[U32 val] box is not a subtype of Generator[U32 val] val: box is not a subcap of val
        : Generator[U32]
          ^
    /home/travis/build/mfelsche/ponycheck/ponycheck/generator.pony:832:24: (USize val, Generator[U32 val] box) is not a pairwise subtype of (USize val, this->Generator[U32 val] box)
              (range_1_size, range_1)
                           ^
```

The actual site of this error is here: https://github.com/mfelsche/ponycheck/blob/shrink/ponycheck/generator.pony#L831

```pony
frequency[U32]([
          (range_1_size, range_1)
          (range_2_size, range_2)
        ])?
```

The definition of [`frequency`](https://github.com/mfelsche/ponycheck/blob/shrink/ponycheck/generator.pony#L244):

```pony
fun frequency[T](
    weighted_generators: ReadSeq[WeightedGenerator[T]])
    : Generator[T] ?
  =>
```

The definition of [`WeightedGenerator`](https://github.com/mfelsche/ponycheck/blob/shrink/ponycheck/generator.pony#L62):

```pony
type WeightedGenerator[T] is (USize, Generator[T] box)
```
I tried to reproduce this in a minimal example, but was not able to. The only result i could produce was #2232 .

This is related to #2232 and has most likely the same cause as it was also introduced with 0.19.1.

Trying to fix this issue by using `[as WeightedGenerator[U32]]`:

```pony
frequency[U32]([as WeightedGenerator[U32]:
  (range_1_size, range_1)
  (range_2_size, range_2)
])?
```

Results in an assertion in reify.c:

```
CC=cc ../ponyc/build/debug/ponyc  --debug ponycheck/test
Building builtin -> /home/mwahl/dev/pony/ponyc/packages/builtin
Building ponycheck/test -> /home/mwahl/dev/pony/ponycheck/ponycheck/test
Building ponytest -> /home/mwahl/dev/pony/ponyc/packages/ponytest
Building time -> /home/mwahl/dev/pony/ponyc/packages/time
Building collections -> /home/mwahl/dev/pony/ponyc/packages/collections
Building .. -> /home/mwahl/dev/pony/ponycheck/ponycheck
Building itertools -> /home/mwahl/dev/pony/ponyc/packages/itertools
Building random -> /home/mwahl/dev/pony/ponyc/packages/random
Generating
 Reachability
src/libponyc/type/reify.c:198: reify: Assertion `(ast_id(typeargs) == TK_TYPEARGS) || (ast_id(typeargs) == TK_NONE)` failed.

Backtrace:
  ../ponyc/build/debug/ponyc(ponyint_assert_fail+0xe9) [0x781c71]
  ../ponyc/build/debug/ponyc(reify+0xa7) [0x6c4ad0]
  ../ponyc/build/debug/ponyc() [0x7569ef]
  ../ponyc/build/debug/ponyc() [0x75a367]
  ../ponyc/build/debug/ponyc() [0x758d58]
  ../ponyc/build/debug/ponyc() [0x758e3b]
  ../ponyc/build/debug/ponyc() [0x759de3]
  ../ponyc/build/debug/ponyc() [0x75a2b8]
  ../ponyc/build/debug/ponyc() [0x75a64c]
  ../ponyc/build/debug/ponyc(reach+0x4e) [0x75a76f]
  ../ponyc/build/debug/ponyc(genexe+0x168) [0x6e1749]
  ../ponyc/build/debug/ponyc(codegen+0x152) [0x6ca237]
  ../ponyc/build/debug/ponyc(generate_passes+0x35) [0x7063e2]
  ../ponyc/build/debug/ponyc() [0x6a08c5]
  ../ponyc/build/debug/ponyc(main+0x4cd) [0x6a0da7]
  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0) [0x7f2281f3f830]
  ../ponyc/build/debug/ponyc(_start+0x29) [0x6a06a9]
Aborted (core dumped)
Makefile:28: recipe for target 'test' failed
make: *** [test] Error 134
```

--- a/src/libponyc/expr/array.c
+++ b/src/libponyc/expr/array.c
@@ -30,6 +30,25 @@ static ast_t* build_array_type(pass_opt_t* opt, ast_t* scope, ast_t* elem_type,
   return array_type;
 }
 
+static ast_t* strip_this_arrow(pass_opt_t* opt, ast_t* ast)
+{
+  if((ast_id(ast) == TK_ARROW) &&
+    (ast_id(ast_child(ast)) == TK_THISTYPE))
+    return ast_childidx(ast, 1);
+
+  if(ast_id(ast) == TK_TUPLETYPE)
+  {
+    ast_t* new_ast = ast_from(ast, TK_TUPLETYPE);
+
+    for(ast_t* c = ast_child(ast); c != NULL; c = ast_sibling(c))
+      ast_append(new_ast, strip_this_arrow(opt, c));
+
+    return new_ast;
+  }
+
+  return ast;
+}
+
 static ast_t* detect_apply_element_type(pass_opt_t* opt, ast_t* ast, ast_t* def)
 {
   // The interface must have an apply method for us to find it.
@@ -59,11 +78,7 @@ static ast_t* detect_apply_element_type(pass_opt_t* opt, ast_t* ast, ast_t* def)
   if(ast_id(typeparams) == TK_TYPEPARAMS)
     elem_type = reify(elem_type, typeparams, typeargs, opt, true);
 
-  if((ast_id(elem_type) == TK_ARROW) &&
-    (ast_id(ast_child(elem_type)) == TK_THISTYPE))
-    elem_type = ast_childidx(elem_type, 1);
-
-  return elem_type;
+  return strip_this_arrow(opt, elem_type);
 }
 
 static ast_t* detect_values_element_type(pass_opt_t* opt, ast_t* ast,

--- a/test/libponyc/array.cc
+++ b/test/libponyc/array.cc
@@ -268,6 +268,29 @@ TEST_F(ArrayTest, InferFromApplyInterfaceIsoElementNoArrowThis)
 }
 
 
+TEST_F(ArrayTest, InferFromApplyInterfaceBoxElementInTuple)
+{
+  // From issue #2233
+  const char* src =
+    "trait box T\n"
+    "class box C1 is T\n"
+    "  new box create() => None\n"
+    "class box C2 is T\n"
+    "  new box create() => None\n"
+    "class box C3 is T\n"
+    "  new box create() => None\n"
+
+    "interface ArrayishOfTAndU64\n"
+    "  fun apply(index: USize): this->(T, U64) ?\n"
+
+    "primitive Foo\n"
+    "  fun apply(): ArrayishOfTAndU64 =>\n"
+    "    [(C1, 1); (C2, 2); (C3, 3)]";
+
+  TEST_COMPILE(src);
+}
+
+
 TEST_F(ArrayTest, InferFromValuesInterface)
 {
   const char* src =

